### PR TITLE
reimplement module aliasing with esbuild plugin

### LINKED ISSUE #6448: user-defined aliases should take precedence over unenv polyfills
We recently introduced module aliasing in https://github.com/cloudflare/workers-sdk/pull/6167

The `experimental:nodejs_compat_v2` flag introduces unenv polyfills via an esbuild plugin onResolve hook. This plugin takes precedence over the user-defined aliases.

A simple repro can be found here https://github.com/RamIdeas/esbuild-alias-wrangler-repro/blob/main/build.mjs

I believe the solution is to reimplement the user-defined aliases via our own plugin which needs to be earlier in the plugin array. Creating this issue to track.

### PR DESCRIPTION
**Pull request by RamIdeas**
## What this PR solves / how to test

Fixes #6448

## Author has addressed the following

- Tests
  - [ ] TODO (before merge)
  - [x] Included
  - [ ] Not necessary because:
- E2E Tests CI Job required? (Use "e2e" label or ask maintainer to run separately)
  - [ ] I don't know
  - [ ] Required / Maybe required
  - [x] Not required because:
- Changeset ([Changeset guidelines](https://github.com/cloudflare/workers-sdk/blob/main/CONTRIBUTING.md#changesets))
  - [ ] TODO (before merge)
  - [x] Included
  - [ ] Not necessary because:
- Public documentation
  - [ ] TODO (before merge)
  - [ ] Cloudflare docs PR(s): 
  - [ ] Not necessary because:

### CODE REVIEW

File: .changeset/tidy-berries-type.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+"wrangler": patch
+---
+
+feat: reimplement module aliasing so user-defined aliases take precedence over other plugins (eg unenv node.js polyfills)
```
Comment (petebacondarwin) [inline .changeset/tidy-berries-type.md:5]:
```suggestion
fix: reimplement module aliasing so user-defined aliases take precedence over other plugins (eg unenv node.js polyfills)
```

File: packages/wrangler/src/deployment-bundle/bundle.ts [hunk 2]
```diff
@@ -278,6 +287,42 @@ export async function bundleWorker(
...
+
+			// reimplement module aliasing as an esbuild plugin onResolve hook
+			build.onResolve({ filter }, (args) => {
+				const aliasPath = alias[args.path];
+				if (aliasPath) {
+					return {
+						// resolve with node resolution
+						path: require.resolve(aliasPath, {
+							// From the esbuild alias docs: "Note that when an import path is substituted using an alias, the resulting import path is resolved in the working directory instead of in the directory containing the source file with the import path."
+							// https://esbuild.github.io/api/#alias:~:text=Note%20that%20when%20an%20import%20path%20is%20substituted%20using%20an%20alias%2C%20the%20resulting%20import%20path%20is%20resolved%20in%20the%20working%20directory%20instead%20of%20in%20the%20directory%20containing%20the%20source%20file%20with%20the%20import%20path.
```
Comment (CarmenPopoviciu) [inline packages/wrangler/src/deployment-bundle/bundle.ts:317]: nit: perhaps just link to https://esbuild.github.io/api/#alias? We already have the relevant text in the comment above this one

File: packages/wrangler/src/deployment-bundle/bundle.ts [hunk 3]
```diff
@@ -278,6 +287,42 @@ export async function bundleWorker(
...
+				return;
+			}
+
+			// filter the hook calls to only those that match the alias keys
+			// this should avoid slowing down builds which don't use aliasing
+			const filter = new RegExp(
+				Object.entries(alias)
+					.map(([key, _value]) => escapeRegex(key))
+					.join("|")
+			);
```
Comment (petebacondarwin) [inline packages/wrangler/src/deployment-bundle/bundle.ts:307]: If we don't need the values then can we just use:

```suggestion
				Object.keys(alias)
					.map((key) => escapeRegex(key))
					.join("|")
			);
```

**Review by petebacondarwin — APPROVED** [review]

File: packages/wrangler/src/deployment-bundle/bundle.ts [hunk 4]
```diff
@@ -317,12 +362,12 @@ export async function bundleWorker(
 				...define,
 			},
 		}),
-		alias,
 		loader: {
 			...COMMON_ESBUILD_OPTIONS.loader,
 			...(loader || {}),
 		},
 		plugins: [
+			aliasPlugin,
```
Comment (CarmenPopoviciu) [inline packages/wrangler/src/deployment-bundle/bundle.ts:370]: what's the order in which these plugins are executed? Is first to last in the `plugins` array, or the other way around? I can't find that mentioned in the esbuild docs

Comment (RamIdeas) [reply packages/wrangler/src/deployment-bundle/bundle.ts:370]: First to last. I confirmed by playing with my repro linked in the issue.

Comment (CarmenPopoviciu) [reply packages/wrangler/src/deployment-bundle/bundle.ts:370]: in that case, what happens if `aliasPlugin` and the unenv one are trying to alias the same module? In that case, won't unenv override the user alias?

Comment (RamIdeas) [reply packages/wrangler/src/deployment-bundle/bundle.ts:370]: Good question. esbuild runs the onResolve hooks on each plugin in order **until** one does not return undefined. It doesn't run anymore onResolve hooks once it receives a path.

Comment (CarmenPopoviciu) [reply packages/wrangler/src/deployment-bundle/bundle.ts:370]: TIL! thx for clarifying that 🙏

**Review by CarmenPopoviciu — APPROVED** [review]

File: packages/wrangler/src/deployment-bundle/bundle.ts (hunk 4 above)
Comment (CarmenPopoviciu) [reply packages/wrangler/src/deployment-bundle/bundle.ts:370]: not sure if we want to capture this use case in a test, I'll leave that up to you

**Comment by IgorMinar** [comment]
Thank you @RamIdeas!

**Pull request merged**